lib/Graph/Graph.cpp

```diff
@@ -2334,70 +2339,74 @@ static bool isGraphStorageNode(Node *n, const Function *F) {
-    llvm::errs() << "The var with name '" << V->getName()
-                 << "' conflicts with a previous definition:\n";
-    llvm::errs() << "Current definition: " << V->getDebugDesc() << "\n";
-    llvm::errs() << "Previous definition: "
-                 << NameToNode[V->getName()]->getDebugDesc() << "\n";
-    dump();
-    llvm_unreachable("Multiple nodes with the same name");
+    msg << "The node with name '" << node.getName()
+        << "' conflicts with a previous definition:\n";
+    msg << "Current definition: " << node.getDebugDesc() << "\n";
+    msg << "Previous definition: "
+        << nameToNode[node.getName()]->getDebugDesc();
+    report(msg.str().c_str());
+    return false;
   }
+  return true;
+}
 
-  NameToNode.clear();
+bool Function::verify() const {
+  bool isValid = true;
+  std::unordered_map<std::string, const Node *> nameToNode;
 
-  for (auto &N : nodes_) {
-    if (NameToNode.insert({N.getName(), &N}).second)
-      continue;
-    /// Output extra information helping to find the error.
-    llvm::outs() << "The node with name '" << N.getName()
-                 << "' conflicts with a previous definition:\n";
-    llvm::errs() << "Current definition: " << N.getDebugDesc() << "\n";
-    llvm::errs() << "Previous definition: "
-                 << NameToNode[N.getName()]->getDebugDesc() << "\n";
-    dump();
-    llvm_unreachable("Multiple nodes with the same name");
+  for (auto *V : getParent()->getConstants()) {
+    isValid &= isUnique(nameToNode, *V, *this);
+  }
+
+  nameToNode.clear();
+  for (const auto &N : nodes_) {
+    isValid &= isUnique(nameToNode, N, *this);
   }
 
   // Any node referenced by one of the graph nodes should be part of the Graph.
   for (const auto &N : nodes_) {
     for (size_t idx = 0, e = N.getNumInputs(); idx < e; ++idx) {
       auto &input = N.getNthInput(idx);
-      (void)input;
       // Verify each input of N.
-      verifyNodeInput(N, idx);
+      isValid &= verifyNodeInput(N, idx);
       bool foundNode =
           std::find(nodes_.begin(), nodes_.end(), *input) != nodes_.end();
-      (void)foundNode;
-      (void)isGraphStorageNode;
-      assert((foundNode || isGraphStorageNode(input, this)) &&
-             "Every node referenced by one of the graph nodes should be part of"
-             "the graph");
+      isValid &= expectCompareTrue(
+          "Every node referenced by one of the graph nodes should be part of "
+          "the graph",
+          foundNode || isGraphStorageNode(input, this), true, &N);
     }
   }
 
   // Check that all uses of each node refer to this node.
   for (const auto &N : nodes_) {
     for (const auto &U : N.getUsers()) {
-      (void)U;
-      assert(U.get()->getNode() == &N &&
-             "All uses of a node should refer to this node");
+      isValid &= expectCompareTrue(
+          "All uses of a node should refer to this node",
+          static_cast<const Node *>(U.get()->getNode()), &N, &N);
```

Comment:
Because the template inference is silly :).
Both this argument and &N have to have the same type, but the template inference sees them as different as one is const and the other non-const.

Arguably it is weird that a const use can return a non-const node.

Reply:
> Arguably it is weird that a const use can return a non-const node.

would it break anything to return const node for const use?

Reply:
if this does not work, we might use `const_cast` to explicitly show intention.

Reply:
I have to check what would imply.
I’ll double check.

On the const_cast thing, i am not fan of removing the constness since using const object here make sense.
Anyway, let us see if the const can be kept on the get thing.

Reply:
Ok, changing the getter to return a const breaks a bunch of thing (nothing we cannot cope, but too many changes for my liking).
I took a different approach and just specified the template argument instead of helping the template inference.
@rdzhabarov let me know if that works for you.